### PR TITLE
Dataset project Id Feature implemented

### PR DESCRIPTION
**Pull request by shivamVarCS**
Design Document:
https://docs.google.com/document/d/1KGM_5uexV5aKqdskPfJo5HzHJ7B8zTBXsHB1OFqxpZY/edit#heading=h.v1bttlen2xl8

### CODE REVIEW

**Comment by sahusanket** [comment]
Please add some description for context. Or attach a JIRA to this PR. 

Need more context on intended behavior projectId vs datasetProjectId in order to review this.

File: src/main/java/io/cdap/delta/bigquery/BigQueryTarget.java [hunk 1]
```diff
@@ -98,19 +106,23 @@ public void configure(Configurer configurer) {
 
   @Override
   public void initialize(DeltaTargetContext context) throws Exception {
+
     Credentials credentials = conf.getCredentials();
-    String project = conf.getProject();
+
+    if (conf.getProject() == null) {
+      return;
```
Comment (sahusanket) [inline src/main/java/io/cdap/delta/bigquery/BigQueryTarget.java]: We should ideally throw a proper error like  illegal argument exception  in such case. 

And all such checks  should be inside the method  `getProject()`  to ensure the method is null safe.

File: src/main/java/io/cdap/delta/bigquery/BigQueryTarget.java [hunk 2]
```diff
@@ -98,19 +106,23 @@ public void configure(Configurer configurer) {
...
+    }
+    String project = conf.getDatasetProject();
+    if (project == null) {
+      return;
```
Comment (sahusanket) [inline src/main/java/io/cdap/delta/bigquery/BigQueryTarget.java]: again similar : adding this check throwing exception within `getDatasetProject()` would be better.

File: src/main/java/io/cdap/delta/bigquery/BigQueryTarget.java [hunk 3]
```diff
@@ -147,22 +159,23 @@ public void initialize(DeltaTargetContext context) throws Exception {
   @Override
   public EventConsumer createConsumer(DeltaTargetContext context) throws IOException {
     Credentials credentials = conf.getCredentials();
-    String project = conf.getProject();
+    if (conf.getProject() == null) {
+      throw new RuntimeException("Project id is not Present");
```
Comment (sahusanket) [inline src/main/java/io/cdap/delta/bigquery/BigQueryTarget.java]: similar : move this inside the `getProject()` , will reduce code repetition.

File: src/main/java/io/cdap/delta/bigquery/BigQueryTarget.java [hunk 4]
```diff
@@ -147,22 +159,23 @@ public void initialize(DeltaTargetContext context) throws Exception {
...
+    }
+    String project = conf.getDatasetProject();
+    if (project == null) {
+      throw new RuntimeException("Project Dataset id is not Present");
```
Comment (sahusanket) [inline src/main/java/io/cdap/delta/bigquery/BigQueryTarget.java]: similar : move this inside the `getDatasetProject()` , will reduce code repetition. 

Also it's better to throw `IllegalArgumentException`

File: src/main/java/io/cdap/delta/bigquery/BigQueryTarget.java [hunk 5]
```diff
@@ -352,5 +395,22 @@ private Credentials getCredentials() throws IOException {
           .createScoped(Collections.singleton("https://www.googleapis.com/auth/cloud-platform"));
       }
     }
+    public String getDatasetProject() {
+      // if it's "auto-detect" that means we need to detect the default project settings
+      // for sandbox you can use `gcloud config set project my-project-id" to set it
```
Comment (sahusanket) [inline src/main/java/io/cdap/delta/bigquery/BigQueryTarget.java]: Please remove or move the `development` or `debugging` related comments to readme.md 

Like here, details about sandbox co-relates to the whole project , not just this method. 

It's usually better to comment only what code is doing.

File: src/test/java/io/cdap/delta/bigquery/BigQueryTargetTest.java [hunk 6]
```diff
@@ -267,7 +268,7 @@ public void testCreateConsumerRetryableFailureForGcsGet() throws Exception {
   }
 
   @Test
-  public void testCreateConsumerNonRetryableFailure() throws Exception {
+    public void testCreateConsumerNonRetryableFailure() throws Exception {
```
Comment (sahusanket) [inline src/test/java/io/cdap/delta/bigquery/BigQueryTargetTest.java]: nit : undo indentation

**Review by sahusanket — CHANGES_REQUESTED** [review]

File: src/main/java/io/cdap/delta/bigquery/BigQueryTarget.java (hunk 1 above)
Comment (shivamVarCS) [reply src/main/java/io/cdap/delta/bigquery/BigQueryTarget.java]: Updated and removed

File: src/main/java/io/cdap/delta/bigquery/BigQueryTarget.java (hunk 2 above)
Comment (shivamVarCS) [reply src/main/java/io/cdap/delta/bigquery/BigQueryTarget.java]: Updated

File: src/main/java/io/cdap/delta/bigquery/BigQueryTarget.java (hunk 3 above)
Comment (shivamVarCS) [reply src/main/java/io/cdap/delta/bigquery/BigQueryTarget.java]: Updated and Removed

File: src/main/java/io/cdap/delta/bigquery/BigQueryTarget.java (hunk 4 above)
Comment (shivamVarCS) [reply src/main/java/io/cdap/delta/bigquery/BigQueryTarget.java]: Updated and removed

File: src/main/java/io/cdap/delta/bigquery/BigQueryTarget.java (hunk 5 above)
Comment (shivamVarCS) [reply src/main/java/io/cdap/delta/bigquery/BigQueryTarget.java]: Updated and Added in the .md file

File: src/test/java/io/cdap/delta/bigquery/BigQueryTargetTest.java (hunk 6 above)
Comment (shivamVarCS) [reply src/test/java/io/cdap/delta/bigquery/BigQueryTargetTest.java]: Updated

File: docs/bigquery-cdcTarget.md [hunk 7]
```diff
@@ -37,8 +37,17 @@ on the new primary key.
...
 
-**Project ID**: Project of the BigQuery dataset. When running on a Dataproc cluster, this can be left blank,
-which will use the project of the cluster.
+**Project ID**: Google Cloud Project ID, which uniquely identifies a project.
+It can be found on the Dashboard in the Google Cloud Platform Console. This is the project
+that the BigQuery job will run in. `BigQuery Job User` role on this project must be granted to the specified service
+account to run the job. If a temporary bucket needs to be created, the bucket will also be created in this project and
+'GCE Storage Bucket Admin' role on this project must be granted to the specified service account to create buckets.
+
+**Dataset Project ID**: Project the dataset belongs to. This is only required if the dataset is not
```
Comment (sahusanket) [inline docs/bigquery-cdcTarget.md]: Please add "destination dataset" to be more clear

File: widgets/bigquery-cdcTarget.json [hunk 8]
```diff
@@ -14,6 +14,14 @@
           "widget-attributes": {
             "default": "auto-detect"
           }
+        },
+        {
+          "name": "datasetProject",
+          "label": "DataSet Project ID",
+          "widget-type": "textbox",
+          "widget-attributes": {
+            "placeholder": "Project the dataset belongs to, if different from the project ID."
```
Comment (sahusanket) [inline widgets/bigquery-cdcTarget.json]: similar : Please add "destination dataset" to be more clear

**Review by sahusanket — CHANGES_REQUESTED** [review]

**Comment by sahusanket** [comment]
I tested the changes in my local, 
The BQ job runs in `DataSet Project ID`  and not in `projectId`

**Comment by shivamVarCS** [comment]
> I tested the changes in my local, The BQ job runs in `DataSet Project ID` and not in `projectId`

I am currently working on fix.

File: docs/bigquery-cdcTarget.md (hunk 7 above)
Comment (shivamVarCS) [reply docs/bigquery-cdcTarget.md]: Updated

File: widgets/bigquery-cdcTarget.json (hunk 8 above)
Comment (shivamVarCS) [reply widgets/bigquery-cdcTarget.json]: Updated

**Review by sahusanket — APPROVED** [review]

**Pull request merged**